### PR TITLE
chore(pypi): publish `git-cliff` on PyPI

### LINKED ISSUE #150: Adding `git-cliff` to PyPI
**Is your feature request related to a problem? Please describe.**
This is similar in scope to the feature described in #133 and implemented in #134. This would make it seamless for Python projects to install and use `git-cliff` and would pair nicely with the newly added support for configuration in `pyproject.toml`. It also fits nicely in the recent [rustification of python tooling trend](https://notes.crmarsh.com/python-tooling-could-be-much-much-faster).

**Describe the solution you'd like**
The approach you described in your [blog post](https://blog.orhun.dev/packaging-rust-for-npm/) for npm can be adapted for PyPI. I think the best approach would be to use [`maturin`](https://github.com/PyO3/maturin#python-packaging-basics). For an example, you can check out [this workflow in the ruff project](https://github.com/charliermarsh/ruff/blob/main/.github/workflows/ruff.yaml).

I'm willing to take a shot at a PR if you're interested in adding this feature. Thanks!

### PR DESCRIPTION
**Pull request by radusuciu**
## Description

This PR updates the CI/CD for packaging git-cliff for PyPI.

## Motivation and Context

Closes #150

## How Has This Been Tested?

At time of writing this has not been tested.

## Types of Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (no code change)
- [ ] Refactor (refactoring production code)
- [x] Other: packaging 

## Checklist:

- [x] My code follows the code style of this project.
- [x] I have updated the documentation accordingly.
- [ ] I have formatted the code with [rustfmt](https://github.com/rust-lang/rustfmt).
- [ ] I checked the lints with [clippy](https://github.com/rust-lang/rust-clippy).
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by radusuciu** [comment]
@orhun Took a crack at this mostly following the example of [`typos`](https://github.com/crate-ci/typos) which you linked in the related issue, modifying the approach to conform to the existing style in the workflow. Please let me know if you've spotted an error or would like to see any changes.

I've set this as a draft since this was an initial effort - I haven't really reviewed everything.

File: .github/workflows/cd.yml [hunk 1]
```diff
@@ -230,6 +242,41 @@ jobs:
...
         env:
           NODE_AUTH_TOKEN: ${{ secrets.NPM_TOKEN }}
+      - name: Build wheels (linux)
+        if: matrix.build.PYPI_PUBLISH == true && matrix.build.NAME.startsWith('linux-')
+        uses: PyO3/maturin-action@v1
+        with:
+          target: ${{ matrix.build.TARGET }}
+          args: --manifest-path git-cliff/Cargo.toml --release --out dist
+          sccache: 'true'
+          manylinux: auto
```
Comment (orhun) [inline .github/workflows/cd.yml:252]: What does this `manylinux` option specify? I'm not so familiar with Python packaging stuff :/

File: .github/workflows/cd.yml [hunk 2]
```diff
@@ -230,6 +242,41 @@ jobs:
...
+          target: ${{ matrix.build.TARGET }}
+          args: --manifest-path git-cliff/Cargo.toml --release --out dist
+          sccache: 'true'
+      - name: Build wheels (windows)
+        if: matrix.build.PYPI_PUBLISH == true && matrix.build.OS.startsWith('windows-')
+        uses: PyO3/maturin-action@v1
+        with:
+          target: ${{ matrix.build.TARGET }}
+          args: --manifest-path git-cliff/Cargo.toml --release --out dist
+          sccache: 'true'
```
Comment (orhun) [inline .github/workflows/cd.yml:266]: These two steps (`Build wheels (macos)` / `Build wheels (windows)`) look identical so maybe merge them if we're not specifying anything that differs? What do you think?

File: .github/workflows/cd.yml [hunk 3]
```diff
@@ -230,6 +242,41 @@ jobs:
...
+      - name: Build wheels (musl)
+        if: matrix.build.PYPI_PUBLISH == true && matrix.build.OS.endsWith('musl-')
```
Comment (orhun) [inline .github/workflows/cd.yml:268]: I think this should be `matrix.build.OS.endsWith('musl')` (mind the "-")

File: .github/workflows/cd.yml [hunk 4]
```diff
@@ -254,6 +301,39 @@ jobs:
...
+      - name: Upload sdist
+        uses: actions/upload-artifact@v3
+        with:
+          name: wheels
+          path: dist
+
+  publish-pypi:
+    name: Publish PyPI packaage
+    runs-on: ubuntu-22.04
+    needs: [pypi-sdist]
```
Comment (orhun) [inline .github/workflows/cd.yml:324]: I think this also needs `publish-binaries`

**Comment by radusuciu** [comment]
Thanks for feedback/corrections! I'll have some time to take a look and incorporate your suggestions this weekend most likely.

**Comment by radusuciu** [comment]
One more general question I had: in `typos` PyPI packaging is [handled in a separate workflow yaml](https://github.com/crate-ci/typos/blob/78a3c66d00d1258d8308756c0e168c44b9ddcdae/.github/workflows/maturin.yml). For this PR I integrated it directly into `cd.yaml` since that seemed to me to be more appropriate given how the workflow is currently structured and I thought would lead to the least amount of duplication. Do you agree with that decision or would you prefer refactoring to a separate file?

**Comment by radusuciu** [comment]
> Great, looking forward to it! For testing purposes on another repository, is there a dry-run option available? Otherwise we can simply create RC releases I think.

Yes! I think we should be able to use TestPyPI: https://packaging.python.org/en/latest/guides/using-testpypi/

**Comment by orhun** [comment]
> One more general question I had: in `typos` PyPI packaging is [handled in a separate workflow yaml](https://github.com/crate-ci/typos/blob/78a3c66d00d1258d8308756c0e168c44b9ddcdae/.github/workflows/maturin.yml). For this PR I integrated it directly into `cd.yaml` since that seemed to me to be more appropriate given how the workflow is currently structured and I thought would lead to the least amount of duplication. Do you agree with that decision or would you prefer refactoring to a separate file?

I think it is good this way. We have the NPM packaging in the same structure as well.

**Comment by orhun** [comment]
> > Great, looking forward to it! For testing purposes on another repository, is there a dry-run option available? Otherwise we can simply create RC releases I think.
> 
> Yes! I think we should be able to use TestPyPI: [packaging.python.org/en/latest/guides/using-testpypi](https://packaging.python.org/en/latest/guides/using-testpypi/).. which I guess isn't fully "dry" in the sense I think you mean, but might be better

Oh, nice. It is definitely better than a dry run.

File: .github/workflows/cd.yml (hunk 1 above)
Comment (radusuciu) [reply .github/workflows/cd.yml:252]: I haven't used it before (or if I have, I have forgotten), I mostly cribbed from other workflow examples. However, here's some info: [`manylinux`](https://github.com/pypa/manylinux) is a [platform compatibility tag](https://packaging.python.org/en/latest/specifications/platform-compatibility-tags/) for python packages.

From the project page:

> The various manylinux tags allow projects to distribute wheels that are automatically installed (and work!) on the vast majority of desktop and server Linux distributions.

File: .github/workflows/cd.yml (hunk 2 above)
Comment (radusuciu) [reply .github/workflows/cd.yml:266]: Agreed, I don't see any issue with combining - can always split back out at a later date if there are platform specific differences to address.

File: .github/workflows/cd.yml (hunk 3 above)
Comment (radusuciu) [reply .github/workflows/cd.yml]: Good catch!

File: .github/workflows/cd.yml (hunk 4 above)
Comment (radusuciu) [reply .github/workflows/cd.yml]: Yes, good catch!

File: .github/workflows/cd.yml [hunk 5]
```diff
@@ -254,6 +301,39 @@ jobs:
...
+      - name: Build sdist
+        uses: PyO3/maturin-action@v1
+        with:
+          command: sdist
+          args: --manifest-path git-cliff/Cargo.toml --out dist
+      - name: Upload sdist
+        uses: actions/upload-artifact@v3
+        with:
+          name: wheels
+          path: dist
```
[the remark this answers is not in the document]
Comment (radusuciu) [reply .github/workflows/cd.yml:319]: `pypi-sdist` builds a source distribution for the project which as the name implies includes the source code... perhaps this doesn't make sense for this project. Let me think about that.

`publish-pypi` handles uploading artifacts to the package index

**Comment by radusuciu** [comment]
Replied to some comments, there are two items to review still. I think we need to test it out against `TestPyPI` next

File: .github/workflows/cd.yml (hunk 1 above)
Comment (orhun) [reply .github/workflows/cd.yml:252]: Got it, thanks for the info!

File: .github/workflows/cd.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (orhun) [reply .github/workflows/cd.yml:319]: hmm, `typos` has source distribution as well, let's leave it in!

**Comment by orhun** [comment]
Looks good for testing now, how do we proceed?

**Comment by radusuciu** [comment]
You'll want to make an account with [`PyPI`](https://pypi.org/) and [`TestPyPI`](https://test.pypi.org/) -- they are separate, and [generate API tokens](https://pypi.org/help/#apitoken) that you can set in the repository secrets under the keys `PYPI_API_TOKEN` and `TESTPYPI_API_TOKEN`.

To roll it out, I can think of two options:

**Option 1**
- Configure workflow to use `TestPyPI` and `TESTPYPI_API_TOKEN` (I need to look this up, but it should be an argument or envvar that is set on the maturin upload action)
- Run workflow
- Verify that everything works
- Re-configure workflow to use `PyPI` and `PYPI_API_TOKEN`

**Option 2**
We tweak the workflow to use either `TestPyPI` or `PyPI` depending on the value of another repository variable eg. `USE_TEST_PYPI`. This is cleaner because it doesn't require editing the workflow to switch between the test and real package index.

If it doesn't work first shot, let's do some testing locally since it'll be faster to iterate than via Actions.

File: .github/workflows/cd.yml (hunk 1 above)
Comment (radusuciu) [reply .github/workflows/cd.yml:252]: Here's a bit more info: https://github.com/PyO3/maturin#manylinux-and-auditwheel

**Comment by radusuciu** [comment]
Seems the `maturin upload` command can be pointed to either `PyPI` or `TestPyPI` by adding an `--repository <REPOSITORY>` argument, with valid values being `pypi` or `testpipy`. Alternatively, the same values will be taken from the `MATURIN_REPOSITORY` environment variable. So maybe something like this:

```yaml
publish-pypi:
  name: Publish PyPI packaage
  runs-on: ubuntu-22.04
  needs: [pypi-sdist, publish-binaries]
  steps:
    - uses: actions/download-artifact@v3
      with:
        name: wheels
    - name: Publish to PyPI
      uses: PyO3/maturin-action@v1
      env:
        MATURIN_PYPI_TOKEN: ${{ secrets.PYPI_API_TOKEN }}
        MATURIN_REPOSITORY: ${{ vars.USE_TEST_PYPI == 'true' && 'testpypi' || 'pypi' }}
      with:
        command: upload
        args: --skip-existing *
```

Also, just noted a spelling mistake: "packaage"

I haven't used `vars` in a workflow before, and [according to the docs, they are still a beta feature](https://docs.github.com/en/actions/learn-github-actions/variables#defining-configuration-variables-for-multiple-workflows), so we can use `secrets` if you prefer.

**Comment by radusuciu** [comment]
> In that case, we need to use PYPI_API_TOKEN or TESTPYPI_API_TOKEN respectfully as well, right?

Yes, you're right, like so:

```yaml
      env:
        MATURIN_PYPI_TOKEN: ${{ vars.USE_TESTPYPI == 'true' && secrets.TESTPYPI_API_TOKEN || secrets.PYPI_API_TOKEN }}
        MATURIN_REPOSITORY: ${{ vars.USE_TESTPYPI == 'true' && 'testpypi' || 'pypi' }}
```

note I also suggest renaming the `USE_TEST_PYPI` repository variable to `USE_TESTPYPI` for consistency/correctness.

**Comment by radusuciu** [comment]
I can commit something when I'm done work, but please feel free to take over if you'd prefer

**Comment by radusuciu** [comment]
Alright, I think we're ready for a test run so long as all of the repository variables/secrets are set as discussed above and defined in `cd.yaml`. Let me know if you'd want me to squash commits.

Also, will running this workflow manually cause any issues or side-effects? We're good on `PyPI` thanks to the test repo but not sure about other bits..

**Comment by radusuciu** [comment]
Just want to note that when I tested locally just now the build command yields an error.

Not sure if this will also be a problem in the context of the [`PyO3/maturin-action@v1`](https://github.com/PyO3/maturin-action) action, but I'm guessing it would be.

```bash
$ maturin build --manifest-path git-cliff/Cargo.toml --release --out dist
📦 Including license file "/home/radu/github/git-cliff/LICENSE"
💥 maturin failed
  Caused by: Found a directory with the module name (git-cliff) next to Cargo.toml, which indicates a mixed python/rust project, but the directory didn't contain an __init__.py file.
```

**Comment by radusuciu** [comment]
I think we're good to go for real now. Remaining questions, which:
- can you run the workflow normally in it's current form without messing up other deployments?
- will the version of the python project be correct? maturin should pull it from `cargo.toml` before deployment, but I saw that the version updates are handled through `release.sh`. I think it should be called but I didn't look at it in depth.
- will maturin run correctly in the context of the action? It runs fine locally now and produced an installable wheel

**Comment by orhun** [comment]
> Alright, I think we're ready for a test run so long as all of the repository variables/secrets are set as discussed above and defined in `cd.yaml`. Let me know if you'd want me to squash commits.

Cool! I will squash the commits on merge so there is no need :3

> Also, will running this workflow manually cause any issues or side-effects? We're good on `PyPI` thanks to the test repo but not sure about other bits..
> can you run the workflow normally in it's current form without messing up other deployments?

`cd.yml` affects the other releases so I was thinking of releasing a "rc" release for testing this. Wouldn't that work? 

> will the version of the python project be correct? maturin should pull it from cargo.toml before deployment, but I saw that the version updates are handled through release.sh. I think it should be called but I didn't look at it in depth.

If `maturin` uses the version from `Cargo.toml`, that's perfectly fine. I don't see a version set anywhere in the workflow file so we should be good. 

> will maturin run correctly in the context of the action? It runs fine locally now and produced an installable wheel

I guess we will see 🐻 

---

I will take a look at the code and add a review. But before actually testing this with a new release, I would like to release `1.2.1` first for the pending changes in `main`. Let's have a clean slate for PyPI :3

From the semver standpoint, would you say we need to bump _minor_ for these changes since there will be a new package?

**Comment by radusuciu** [comment]
> `cd.yml` affects the other releases so I was thinking of releasing a "rc" release for testing this. Wouldn't that work?

That should work with the downside that you're creating release candidates for something that isn't a core feature -- but because they're rc's it should be fine and not annoying to end-users? Maybe we're splitting hairs here lol, but here are some alternatives I can think of:
- split off this workflow file and attempt to avoid code duplication through use of [reusable workflows](https://docs.github.com/en/actions/using-workflows/reusing-workflows).
- split off workflow file into separate repository that only exists to test the workflow with TestPyPi... maybe can even do this with my fork
- maybe try using [`act`](https://github.com/nektos/act) locally? Since I think we just need to see that this workflow works with `TestPyPi`, you can temporarily edit (or copy) `cd.yaml` to remove `cargo` and `npm` publishing steps

> I will take a look at the code and add a review. But before actually testing this with a new release, I would like to release `1.2.1` first for the pending changes in `main`. Let's have a clean slate for PyPI :3

Sounds good! Now that you mention it, it should be possible to test this workflow without consequence at least once with the release of `1.2.1`. If you merge this PR first, the repository variable `USE_TESTPYPI` is `true` and `TESTPYPI_API_TOKEN` is set, and then you release `1.2.1`, we should see a `git-cliff==1.2.1` on TestPyPi - which will not pollute the eventual PyPi release.

> From the semver standpoint, would you say we need to bump _minor_ for these changes since there will be a new package?

I would agree with that!

**Comment by radusuciu** [comment]
Hey I started testing this in a separate branch and found this error: `"working-directory" is not available with "uses". it is only available with "run"` (message as reported by `actionlint`).

**Comment by radusuciu** [comment]
Alright, so I hacked away parts of the CD workflow until it got to the bits that are relevant to this PR (I jettisoned things like signing, npm publishing etc. that needed more set up in my forked repo).

Now failing at:

```
Run actions/download-artifact@v3
  with:
    name: wheels
Starting download for wheels
Error: Unable to find any artifacts for the associated workflow
```

See: https://github.com/radusuciu/git-cliff/actions/runs/5106471991/jobs/9178781857

I'll continue on it but just wanted to leave an update.

**Comment by radusuciu** [comment]
So I think I've sorted out all of the issues with paths that were causing problems. I still haven't managed to get a workflow to publish to TestPyPI though. Here's the latest:

https://github.com/radusuciu/git-cliff/actions/runs/5116808835/jobs/9199428899

```
##[debug]followSymbolicLinks 'true'
##[debug]implicitDescendants 'true'
##[debug]matchDirectories 'true'
##[debug]omitBrokenSymbolicLinks 'true'
##[debug]Search path '/home/runner/work/git-cliff/git-cliff/pypi/wheels'
/home/runner/work/_temp/d5019b40-47ac-4166-bc22-d10defe06b01/maturin upload --skip-existing /home/runner/work/git-cliff/git-cliff/pypi/wheels/test_pypi_deploy_git_cliff-1.3.1-py3-none-manylinux_2_31_x86_64.whl /home/runner/work/git-cliff/git-cliff/pypi/wheels/test_pypi_deploy_git_cliff-1.3.1.tar.gz /home/runner/work/git-cliff/git-cliff/pypi/wheels/test_pypi_deploy_git_cliff-1.3.14-py3-none-manylinux_2_17_x86_64.manylinux2014_x86_64.whl /home/runner/work/git-cliff/git-cliff/pypi/wheels/test_pypi_deploy_git_cliff-1.3.14.tar.gz
🚀 Uploading 4 packages
💥 maturin failed
  Caused by: 💥 Failed to upload "test_pypi_deploy_git_cliff-1.3.1-py3-none-manylinux_2_31_x86_64.whl" (3.7 MB)
  Caused by: Failed to upload the wheel with status 400: <html>
 <head>
  <title>400 This filename has already been used, use a different version. See https://test.pypi.org/help/#file-name-reuse for more information.</title>
 </head>
 <body>
  <h1>400 This filename has already been used, use a different version. See https://test.pypi.org/help/#file-name-reuse for more information.</h1>
  The server could not comply with the request since it is either malformed or otherwise incorrect.<br/><br/>
This filename has already been used, use a different version. See https://test.pypi.org/help/#file-name-reuse for more information.
```

I incremented versions a bunch while testing and it seems that for some reason we're building wheels for two different versions. I'm planning on wiping all of my testing commits to see if it'll work when starting from a clean slate. It's possible that we're already there.

**Comment by radusuciu** [comment]
Hi @orhun, after a lot of back and forth, mostly dealing with issues related to paths, I've got a successful run of a pared down action: https://github.com/radusuciu/git-cliff/actions/runs/5123788390.

Here's the package on TestPyPI: https://test.pypi.org/project/test-pypi-deploy-git-cliff/

**Review by orhun — APPROVED** [review]
Sorry for the delay and thank you so much for the implementation/testing! 🐻

**Comment by radusuciu** [comment]
No worries!

Please feel free to tag me in if there's any issues

**Comment by orhun** [comment]
Hey @radusuciu, I created the RC release today and there was a build error: https://github.com/orhun/git-cliff/actions/runs/5911302728/job/16033716352

Can you look into it? 🐻

**Comment by orhun** [comment]
Just disabled `win32-arm64` and released another RC, it [worked](https://github.com/orhun/git-cliff/actions/runs/5919256617/job/16049512717) this time! 🥳

- https://test.pypi.org/project/git-cliff/

```
$ pip install -i https://test.pypi.org/simple/ git-cliff
```

**Comment by radusuciu** [comment]
> Hey @radusuciu, I created the RC release today and there was a build error: https://github.com/orhun/git-cliff/actions/runs/5911302728/job/16033716352
> 
> Can you look into it? 🐻
> 
> edit: maybe we can disable `win32-arm64` releases for PyPI...

Hmm, most of the references I've come across for this point towards issues installing `ring` due to something `musl` related but that doesn't make sense here.

I was just travelling but jotting down some links for later (may or may not be relevant, I'll look into it more)
- https://stackoverflow.com/questions/68149955/rust-sorry-unimplemented-64-bit-mode-not-compiled-in/70628855#70628855
- https://github.com/rust-lang/rust/issues/63939#issuecomment-665952545
- https://github.com/briansmith/ring/issues/1488

**Pull request merged**